### ISSUE TITLE
rechunker implementation

### ISSUE THREAD

**Issue by chiaweh2** [NOAA-PSL/ufs2arco#6]
@timothyas thanks for introducing the package!

I was just going to suggest you to use rechunker in the tutorial meeting and notice this https://github.com/NOAA-PSL/ufs2arco/blob/9e9c239697ec0c5d977ddee045559ea67e7371d8/ufs2arco/ufsdataset.py#L156. Since the file on the cloud is in the netcdf format. I don't know if rechunker can do what chunk did. But I do think rechunker would be faster based on my past test. 

The only thing I can think of is to create a zarr store at local first based on the original chunking and rechunk it after. Or if your rechunked ML data in zarr is also open to the public, it would be a way to go! If you are thinking about doing this, I am happy to help on this since this would also be useful when I am rechunking the regional mom6 data in the future. Just that we need to think of a good strategy.

Or maybe have the option of rechunker when the user is opening the dataset in zarr format? Just throwing some thoughts here!

**Comment by timothyas** [comment]
Hi @chiaweh2, thanks for this! I briefly tried to use rechunker, but didn't really see any improvements. I am not really sure why this was, but here is a more detailed outline of what I tried. I had this written up as an issue in a separate repository but it's more useful here

---

I suspect(ed) that integrating rechunker will speed up our current workflow. However, I'm running some tests on GCP that give a more nuanced view. This issue is just meant to document that process.

## 1 Degree Replay Dataset

Everything listed below is relevant to the 1 degree replay dataset, found [here](https://noaa-ufs-gefsv13replay-pds.s3.amazonaws.com/index.html#1deg/). I'm running on GCP using a single c2-standard-60 node (60 vCPUs, 240GB memory, amd64) in zone `us-central1-c`. I'm using the default threaded scheduler with dask. I'm pulling data from s3 and storing to a persistent local storage there (`/contrib` in the parallel works cluster).

### Using default dask `.rechunk`

Accessed by xarray's `.chunk` method results in download, rechunk, and store time of ~1 minute per day (i.e., 15 seconds per DA cycle).

### Using rechunker

With `max_mem=30GB` based on my assumption of ~8 dask workers per node results in ~1 min 40 seconds per day (i.e., 25 seconds per DA cycle)~ actually the timing is pretty variable, likely due to the network, I've seen as low as 10 sec per DA cycle and as high as 27 sec per DA cycle.

Additionally, I cannot seem to properly append zarr stores in time. No matter what I get the following error unless I delete the target store - **not the temp_store**. Despite what the error says, I am very certain that this is being deleted. 

<details>

<summary>The error</summary>

```python
Traceback (most recent call last):
  File "/contrib/Tim.Smith/ufn/scripts/read_from_s3.py", line 38, in <module>
    replay.store_dataset(ds, **kw)
  File "/contrib/Tim.Smith/UFS2ARCO/src/UFS2ARCO/ufsdataset.py", line 211, in store_dataset
    self._store_data_vars(xds, **kwargs)
  File "/contrib/Tim.Smith/UFS2ARCO/src/UFS2ARCO/ufsdataset.py", line 248, in _store_data_vars
    chunk_plan = rechunk(
                 ^^^^^^^^
  File "/contrib/Tim.Smith/miniconda3/envs/ufs2arco/lib/python3.11/site-packages/rechunker/api.py", line 303, in rechunk
    copy_spec, intermediate, target = _setup_rechunk(
                                      ^^^^^^^^^^^^^^^
  File "/contrib/Tim.Smith/miniconda3/envs/ufs2arco/lib/python3.11/site-packages/rechunker/api.py", line 464, in _setup_rechunk
    copy_spec = _setup_array_rechunk(
                ^^^^^^^^^^^^^^^^^^^^^
  File "/contrib/Tim.Smith/miniconda3/envs/ufs2arco/lib/python3.11/site-packages/rechunker/api.py", line 599, in _setup_array_rechunk
    target_array = _zarr_empty(
                   ^^^^^^^^^^^^
  File "/contrib/Tim.Smith/miniconda3/envs/ufs2arco/lib/python3.11/site-packages/rechunker/api.py", line 147, in _zarr_empty
    return store_or_group.empty(
           ^^^^^^^^^^^^^^^^^^^^^
  File "/contrib/Tim.Smith/miniconda3/envs/ufs2arco/lib/python3.11/site-packages/zarr/hierarchy.py", line 1194, in empty
    return self._write_op(self._empty_nosync, name, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/contrib/Tim.Smith/miniconda3/envs/ufs2arco/lib/python3.11/site-packages/zarr/hierarchy.py", line 935, in _write_op
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "/contrib/Tim.Smith/miniconda3/envs/ufs2arco/lib/python3.11/site-packages/zarr/hierarchy.py", line 1200, in _empty_nosync
    return empty(store=self._store, path=path, chunk_store=self._chunk_store, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/contrib/Tim.Smith/miniconda3/envs/ufs2arco/lib/python3.11/site-packages/zarr/creation.py", line 312, in empty
    return create(shape=shape, fill_value=None, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/contrib/Tim.Smith/miniconda3/envs/ufs2arco/lib/python3.11/site-packages/zarr/creation.py", line 203, in create
    init_array(
  File "/contrib/Tim.Smith/miniconda3/envs/ufs2arco/lib/python3.11/site-packages/zarr/storage.py", line 451, in init_array
    _init_array_metadata(
  File "/contrib/Tim.Smith/miniconda3/envs/ufs2arco/lib/python3.11/site-packages/zarr/storage.py", line 521, in _init_array_metadata
    raise ContainsArrayError(path)
zarr.errors.ContainsArrayError: path 'tmp' contains an array
```

</details>

On top of that it is not possible to do this without providing a `temp_store` to rechunker.rechunk, despite the fact that the [documentation](https://rechunker.readthedocs.io/en/latest/api.html#rechunk-function) states that this is an optional argument.


<details>

<summary>Here is my script, note that this has an error when `rmtree(replay.path_out)` is not called each time (which is not what we want)</summary>

```python
import os
import datetime
from shutil import rmtree

from UFS2ARCO import FV3Dataset
from UFS2ARCO.replay import cached_replay_path_1degree as replay_path

from timer import Timer

if __name__ == "__main__":

    walltime = Timer()
    localtime = Timer()

    walltime.start("Initializing job")

    replay = FV3Dataset(path_in=replay_path, config_filename="config-replay.yaml")
    kw = {"mode": "w"}
    year = 1994
    month = 1
    if os.path.isdir(replay.path_out):
        rmtree(replay.path_out)
    for day in range(1,3):
        for hour in [0, 6, 12, 18]:

            date = datetime.datetime(year=year, month=month, day=day, hour=hour)

            if os.path.isdir("../zstores/tmp"):
                rmtree("../zstores/tmp")
            # The code only runs if the following two lines are un-commented... but this is not at all what we want!
            #if os.path.isdir(replay.path_out):
            #    rmtree(replay.path_out)
            print(os.listdir("../zstores"))
            localtime.start(f"Pulling {str(date)}")
            ds = replay.open_dataset(date, fsspec_kwargs={"s3":{"anon": True}}, engine="h5netcdf")
            replay.store_dataset(ds, **kw)
            kw = {"mode": "a", "append_dim":"time"}
            localtime.stop()
```

</details>

So TL; DR I might just roll with the default chunking unless it becomes a problem because it seems to be not too bad, and avoids these issues.

**Comment by chiaweh2** [comment]
Thanks for sharing the test! I think it is highly possible rechunker is not optimal for every single use case (as always). In this case, it does seems using `.chunk` is enough for the package. Just curious, when you get the file from the cloud to do the rechunker test, is the file in zarr or in netcdf (if it is the file in [here](https://noaa-ufs-gefsv13replay-pds.s3.amazonaws.com/index.html#1deg/1994/01/1994010100/), I assume netcdf?) If that is the case, maybe it is also related to that the need of loading all files to memory and do rechunk of xarray object instead of doing rechunk on zarr store chunks? But this is just my wild thoughts on the possible factors.

**Comment by timothyas** [comment]
@chiaweh2 You're totally right, it could be that the 1 degree dataset is just not really suited for rechunker, and maybe the quarter degree data would see more advantages. Also, the chunking scheme that we went with for the dataset is not that different from the original dataset (chunk = single timestep, all 2D and 3D information), so maybe there is little that needs to be improved with rechunker. Lastly it could be that network calls dominate the wait time, so the rechunking is coming out in the wash.

So rechunker could be good to pursue in the future. With that said, I think the better approach for very large datasets would be to go with pangeo-forge, which uses rechunker. So maybe we don't need to work out those details if we figure out how to integrate with pangeo-forge.
